pkg/cri/sbserver/sandbox_stats_windows.go

```diff
@@ -19,17 +19,397 @@ package sbserver
+	"github.com/containerd/typeurl"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
 )
 
-func (c *criService) podSandboxStats(ctx context.Context, sandbox sandboxstore.Sandbox, stats interface{}) (*runtime.PodSandboxStats, error) {
-	return nil, fmt.Errorf("pod sandbox stats not implemented on windows: %w", errdefs.ErrNotImplemented)
+func (c *criService) podSandboxStats(
+	ctx context.Context,
+	sandbox sandboxstore.Sandbox) (*runtime.PodSandboxStats, error) {
+	meta := sandbox.Metadata
+
+	if sandbox.Status.Get().State != sandboxstore.StateReady {
+		return nil, fmt.Errorf("failed to get pod sandbox stats since sandbox container %q is not in ready state", meta.ID)
+	}
+
+	timestamp := time.Now()
+	podSandboxStats := &runtime.PodSandboxStats{
+		Windows: &runtime.WindowsPodSandboxStats{},
+		Attributes: &runtime.PodSandboxAttributes{
+			Id:          meta.ID,
+			Metadata:    meta.Config.GetMetadata(),
+			Labels:      meta.Config.GetLabels(),
+			Annotations: meta.Config.GetAnnotations(),
+		},
+	}
+
+	metrics, containers, err := c.listWindowsMetricsForSandbox(ctx, sandbox)
+	if err != nil {
+		return nil, fmt.Errorf("failed to obtain container stats during podSandboxStats call: %w", err)
+	}
+
+	statsMap, err := convertMetricsToWindowsStats(metrics, sandbox.Config.GetWindows().GetSecurityContext().GetHostProcess())
+	if err != nil {
+		return nil, err
+	}
+
+	podCPU, containerStats, err := c.toPodSandboxStats(sandbox, statsMap, containers, timestamp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert container stats during podSandboxStats call: %w", err)
+	}
+	podSandboxStats.Windows.Cpu = podCPU.Cpu
+	podSandboxStats.Windows.Memory = podCPU.Memory
+	podSandboxStats.Windows.Containers = containerStats
+
+	podSandboxStats.Windows.Network = windowsNetworkUsage(ctx, sandbox, timestamp)
+
+	pidCount, err := c.getSandboxPidCount(ctx, sandbox)
+	if err != nil {
+		return nil, err
+	}
+
+	podSandboxStats.Windows.Process = &runtime.WindowsProcessUsage{
+		Timestamp:    timestamp.UnixNano(),
+		ProcessCount: &runtime.UInt64Value{Value: pidCount},
+	}
+
+	c.saveSandBoxMetrics(podSandboxStats.Attributes.Id, podSandboxStats)
+
+	return podSandboxStats, nil
+}
+
+func convertMetricsToWindowsStats(metrics []*types.Metric, isHostProcess bool) (map[string]*wstats.Statistics, error) {
+	statsMap := make(map[string]*wstats.Statistics)
+	for _, stat := range metrics {
+		containerStatsData, err := typeurl.UnmarshalAny(stat.Data)
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract metrics for container with id %s: %w", stat.ID, err)
+		}
+
+		// extract the metrics if available for this container
+		// containerStatsData can be nil for pods that don't have an actual podsandbox container such as HPC
+		// In the case of HostProcess sandbox container we will use the nil value for the statsmap which is used later
+		containerStats, ok := containerStatsData.(*wstats.Statistics)
+		if !ok && !isHostProcess {
```

Comment:
Does this actually work E2E for HPC? We won't error out here, but it looks like we would add a `nil` `containerStats` to `statsMap`, and then hit the `if containerMetric == nil` case on line 128, and error out there. If the intent is to error out for HPC, probably simpler to just do that here?

Reply:
The sandbox doesn't have this check, (lines 113-120).  Line 128 is for the workload containers which would be an unexpected case.   See https://github.com/containerd/containerd/pull/7099#discussion_r1097992153.  We don't want to error out here in HPC scenario since it is expected.  

This is a small nuance and open to suggestions to make this clearer

Reply:
I see, so we do get valid stats for an HPC container, just not for the pod/pause itself?

In that case, I think we could make this a bit more foolproof by passing in `sandbox` and making this check be `if !ok && !(isHostProcess && stat.ID == sandbox.ID`. Then we would still throw an error if an HPC container didn't have valid stats.

What do you think?

Reply:
> I see, so we do get valid stats for an HPC container, just not for the pod/pause itself?

correct

> In that case, I think we could make this a bit more foolproof by passing in sandbox and making this check be if !ok && !(isHostProcess && stat.ID == sandbox.ID). Then we would still throw an error if an HPC container didn't have valid stats.
>
> What do you think?

I like it, that make is more explicit.  Updated.